File: python/剑指offer/面试题55 - II. 平衡二叉树.py
# 后序遍历 + 减枝
class Solution:
    def isBalanced(self, root: TreeNode) -> bool:
        def recur(root):
            if not root: return 0
            left = recur(root.left)
            if left == -1: return -1
            right = recur(root.right)
            if right == -1: return -1
            return max(left, right) + 1 if abs(left - right) <= 1 else -1

        return recur(root) != -1

# 先序遍历 + 深度

class Solution:
    def isBalanced(self, root: TreeNode) -> bool:
        def depth(root:TreeNode):
            if not root:
                return 0
            return max(depth(root.left),depth(root.right))+1
        if not root:
            return True
        left_depth = depth(root.left)
        right_depth = depth(root.right)
        return abs(left_depth-right_depth)<=1 and self.isBalanced(root.left) and self.isBalanced(root.right)

File: go/剑指offer/面试题55 - II. 平衡二叉树.go
package main

// 后序遍历

func isBalanced(root *TreeNode) bool {
	return recur(root) != -1
}

func recur(root *TreeNode) int {
	if root == nil {
		return 0
	}
	left := recur(root.Left)
	if left == -1 {
		return -1
	}
	right := recur(root.Right)
	if right == -1 {
		return -1
	}
	if abs(left-right) > 1 {
		return -1
	} else {
		return max(left, right) + 1
	}
}

func abs(a int) int {
	if a < 0 {
		return -1 * a
	} else {
		return a
	}
}

func max(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}

// 先序遍历
// func isBalanced(root *TreeNode) bool {
// 	if root == nil {
// 		return true
// 	}
// 	return abs(maxDepth(root.Left)-maxDepth(root.Right)) <= 1 && isBalanced(root.Left) && isBalanced(root.Right)
// }
//
// func depth(root *TreeNode) int {
// 	if root == nil {
// 		return 0
// 	}
// 	return max(depth(root.Left), depth(root.Right)) + 1
// }
//

//
